pkg/trie/triedb/triedb.go

```diff
@@ -267,6 +342,158 @@ func (t *TrieDB) inspect(
 	}
 }
 
+// fix is a helper function to reorganise the nodes after deleting a branch.
+// For example, if the node we are deleting is the only child for a branch node, we can transform that branch in a leaf
+func (t *TrieDB) fix(node Node) (Node, error) {
+	usedIndex := make([]byte, 0)
+
+	switch branch := node.(type) {
+	case Branch:
+		for i := 0; i < codec.ChildrenCapacity; i++ {
+			if branch.children[i] != nil {
+				if len(usedIndex) == 2 {
+					break
+				}
+				usedIndex = append(usedIndex, byte(i))
+			}
+		}
+
+		if len(usedIndex) == 0 {
+			if branch.value == nil {
+				panic("branch with no subvalues. Something went wrong.")
+			}
+
+			// Make it a leaf
+			return Leaf{branch.partialKey, branch.value}, nil
+		} else if len(usedIndex) == 1 && branch.value == nil {
+			// Only one onward node. use child instead
+			idx := usedIndex[0] //nolint:gosec
+			// take child and replace it to nil
+			child := branch.children[idx]
+			branch.children[idx] = nil
+
+			var stored StoredNode
+			switch n := child.(type) {
+			case InMemory:
+				stored = t.storage.destroy(n.idx)
+			case Persisted:
+				handle, err := t.lookupNode(n.hash)
+				if err != nil {
+					return nil, err
+				}
+				stored = t.storage.destroy(handle)
+			}
+
+			var childNode Node
+			switch n := stored.(type) {
+			case NewStoredNode:
+				childNode = n.node
+			case CachedStoredNode:
+				t.deathRow[n.hash] = nil
+				childNode = n.node
+			}
+
+			combinedKey := branch.partialKey
+			combinedKey = append(combinedKey, idx)
+			combinedKey = append(combinedKey, childNode.getPartialKey()...)
+
+			switch n := childNode.(type) {
+			case Leaf:
+				return Leaf{combinedKey, n.value}, nil
+			case Branch:
+				return Branch{combinedKey, n.children, n.value}, nil
+			default:
+				panic("unreachable")
+			}
+		} else {
+			// Restore branch
+			return Branch{branch.partialKey, branch.children, branch.value}, nil
+		}
+	default:
+		panic("fix should be only called with branch nodes")
+	}
+}
+
+// removeInspector removes the key node from the given node `stored`
+func (t *TrieDB) removeInspector(stored Node, keyNibbles []byte, oldValue *nodeValue) (action, error) {
+	partial := keyNibbles
+
+	switch n := stored.(type) {
+	case Empty:
+		return delete{}, nil
+	case Leaf:
+		existingKey := n.partialKey
+
+		if bytes.Equal(existingKey, partial) {
+			// This is the node we are looking for so we delete it
+			t.replaceOldValue(oldValue, n.value)
+			return delete{}, nil
+		}
+		// Wrong partial, so we return the node as is
+		return restore{n}, nil
+	case Branch:
+		if len(partial) == 0 {
+			if n.value == nil {
+				// Nothing to delete since the branch doesn't contains a value
+				return restore{n}, nil
+			}
+			// The branch contains the value so we delete it
+			t.replaceOldValue(oldValue, n.value)
+			newNode, err := t.fix(Branch{n.partialKey, n.children, nil})
+			if err != nil {
+				return nil, err
+			}
+			return replace{newNode}, nil
+		} else {
```

Comment:
can we forego this else condition? Looks like it will return in the `len(partial) == 0` condition.